src/openms/source/CHEMISTRY/NucleicAcidSpectrumGenerator.cpp

```diff
@@ -223,84 +215,90 @@ namespace OpenMS
+    vector<double> fragments_left, fragments_right, thiol_offsets;
     Size start = add_first_prefix_ion_ ? 0 : 1;
-    if ((add_a_ions_ || add_b_ions_ || add_c_ions_ || add_d_ions_ ||
-         add_aB_ions_) && (oligo.size() > start + 1))
+    if ((add_a_ions_ || add_b_ions_ || add_c_ions_ || add_d_ions_ || add_aB_ions_) && (oligo.size() > start + 1))
     {
       fragments_left.resize(oligo.size() - 1);
+      // Drop the final thiol, 'cause its not linking anything
+      thiols.resize(oligo.size() - 1);
       fragments_left[0] = ribo_masses[0] + five_prime_mass;
       for (Size i = 1; i < oligo.size() - 1; ++i)
       {
-        fragments_left[i] = (fragments_left[i - 1] + ribo_masses[i] +
-                             backbone_mass);
+        fragments_left[i] = (fragments_left[i - 1] + ribo_masses[i] + backbone_mass + thiols[i - 1]);
       }
+      // with thiols c and d ions have a 15.99 mass shift, we calculated that above now we add it
+      vector<double> frag_l_thiol(fragments_left.size());
+      std::transform(fragments_left.begin(), fragments_left.end(), thiols.begin(), frag_l_thiol.begin(), std::plus<double>());
+
       if (add_a_ions_)
       {
-        addFragmentPeaks_(spectrum, fragments_left, "a", a_ion_offset,
-                          a_intensity_, start);
+        addFragmentPeaks_(spectrum, fragments_left, "a", a_ion_offset, a_intensity_, start);
       }
       if (add_b_ions_)
       {
-        addFragmentPeaks_(spectrum, fragments_left, "b", b_ion_offset,
-                          b_intensity_, start);
+        addFragmentPeaks_(spectrum, fragments_left, "b", b_ion_offset, b_intensity_, start);
       }
       if (add_c_ions_)
       {
-        addFragmentPeaks_(spectrum, fragments_left, "c", c_ion_offset,
-                          c_intensity_, start);
+        addFragmentPeaks_(spectrum, frag_l_thiol, "c", c_ion_offset, c_intensity_, start);
       }
       if (add_d_ions_)
       {
-        addFragmentPeaks_(spectrum, fragments_left, "d", d_ion_offset,
-                          d_intensity_, start);
+        addFragmentPeaks_(spectrum, frag_l_thiol, "d", d_ion_offset, d_intensity_, start);
       }
       if (add_aB_ions_) // special case
       {
         addAMinusBPeaks_(spectrum, fragments_left, oligo, start);
       }
     }
 
-    if ((add_w_ions_ || add_x_ions_ || add_y_ions_ || add_z_ions_) &&
-        (oligo.size() > 1))
+    if ((add_w_ions_ || add_x_ions_ || add_y_ions_ || add_z_ions_) && (oligo.size() > 1))
     {
       fragments_right.resize(oligo.size() - 1);
       fragments_right[0] = ribo_masses.back() + three_prime_mass;
       for (Size i = 1; i < oligo.size() - 1; ++i)
       {
         Size ribo_index = oligo.size() - i - 1;
-        fragments_right[i] = (fragments_right[i - 1] + ribo_masses[ribo_index] +
-                              backbone_mass);
+        fragments_right[i] = (fragments_right[i - 1] + ribo_masses[ribo_index] + backbone_mass + thiols[ribo_index]);
```

Comment:
there is special treatment of the terminal one in 241. is this also needed here?

Reply:
Nope, we use the same `thiols` object here, so the shortening doesn't need to be redone

Reply:
then I don't get the logic. You check for ion type above and then perform the shortening based on what ions you want to generate. And then there is a separate block that reuses it. what if the first one is never executed?

Reply:
You are correct, in the event that none of the 5' ion types were selected this would cause an error. I'll move the resize out of the if block.